-lオプションの追加

guchi5 opened the pull request:
-lオプションを実装しました。

guchi5 committed:
-lオプションの追加

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,73 @@ def show_files(matrix)
+  show_status += FILE_MODE_TABLE['permission'][status[5].to_i].to_s
+
+  if File.setuid?(absolute_path)
+    special_permission = FILE_MODE_TABLE['special_permission'][status[2].to_i].to_s
```
重複しているので、これは `if` の外に追いてもいいかもしれませんね。
使わない場合もありますが、まあ大した問題ではないかなと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,73 @@ def show_files(matrix)
 end
 
+# ファイルモードの数値を記号表記に変換する
+def convert_file_mode(status, file) # rubocop:disable Metrics/PerceivedComplexity
```
disableにする際は必ず理由を書いてください。
今回はメソッドに分けるなどで対応できるかなと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -17,6 +46,34 @@ def create_matrix(files, max_col_size)
   matrix
 end
 
+# ロングオプション用の表示用行列生成
```
この程度のコメントであればできるだけメソッド名であらわせるようにするといいかなと思います。
`create_detail_matrix_for_long_option` などとすればコメント不要ですよね。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,73 @@ def show_files(matrix)
+
+  show_status += FILE_MODE_TABLE['permission'][status[5].to_i].to_s
+
+  if File.setuid?(absolute_path)
+    special_permission = FILE_MODE_TABLE['special_permission'][status[2].to_i].to_s
+    special_permission.upcase if show_status[3] != 'x'
+    show_status[3] = special_permission
+
+  elsif File.setgid?(absolute_path)
+    special_permission = FILE_MODE_TABLE['special_permission'][status[2].to_i].to_s
+    special_permission.upcase if show_status[6] != 'x'
+    show_status[6] = special_permission
+
+  elsif File.sticky?(absolute_path)
+    special_permission = FILE_MODE_TABLE['special_permission'][status[2].to_i].to_s
+    special_permission.upcase if show_status[9] != 'x'
+    show_status[9] = special_permission
+  end
```
setuid, setgid, sticky bitの対応までやられているのはよいですね :+1: 
` special_permission.upcase` だと、その場で大文字を返しておわりで `special_permission` は変化しないので、これだと大文字表示にはならないかなと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,73 @@ def show_files(matrix)
+    file += " -> #{File.readlink(absolute_path)}" if File.symlink?(absolute_path)
+    [show_status, hard_link_num, user_name, group_name, size, date, file]
+  end
+  col_size = 7
```
ここわざわざcol_sizeを上書きする必要ありますかね？
`show_detail_files(create_detail_matrix(files, col_size))` でしか使われてなさそうなので、`show_detail_files(create_detail_matrix(files, COL_SIZE_FOR_DETAIL))` のようにすればいいだけではと思いました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,73 @@ def show_files(matrix)
-matrix = create_matrix(files, MAX_COL_SIZE)
+total_files_num = 0
+if long_option
+  files.map! do |file|
```
`files` として最初はファイル名の一覧ですが、ここで配列の配列という形の違うものになっているのがトリッキーに感じます。特に上書きする必要はないと思うので、 `file_attributes` など、別の変数としてはどうでしょうか。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -17,6 +46,34 @@ def create_matrix(files, max_col_size)
+def show_detail_files(matrix)
+  matrix[0][:col].length.times do |i|
+    matrix.each_with_index do |value, j|
+      if j == 2 || j == 3 || j == matrix.length - 1
```
`2` や `3` など、埋め込みの数値が多くつかわれていてパッと見た感じだとどこの何をあらわしているのかがわかりにくいです。Hashなどを使ってもうちょっと読みやすくできないか、考えてみてください。

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
可読性のため修正
- show_detail_filesの修正
- create_detail_matrixをcreate_matrix_for_long_optionに変更

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,66 @@ def show_files(matrix)
+  puts "total #{total_files_num}"
+  file_attributes = file_attributes.transpose.flatten
+  show_detail_files(create_matrix_for_long_option(file_attributes, COL_SIZE_FOR_DETAIL))
+  return
```
コマンドを実装する場合、 `return` ではなく `exit` で終了することも多いです。
`exit` ですとコマンド自体が終了していることも明示していることになりますし、ステータスコードなどにも対応できるため、この場合は `exit` のほうが適当かなと思います。
https://docs.ruby-lang.org/ja/latest/method/Kernel/m/exit.html

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -17,6 +47,33 @@ def create_matrix(files, max_col_size)
+  col_detail = { user: 2, group: 3, file: matrix.length - 1 } # 左寄せ表示させる列
+  matrix[0][:col].length.times do |i|
+    matrix.each_with_index do |value, j|
+      if j == col_detail[:user] || j == col_detail[:group] || j == col_detail[:file]
```
`matrix` 自体の各要素をHashにしてしまってもよいかなと思います。
RubyではHashも順序を保持するので、eachでの列挙にも対応できます。

参考: https://docs.ruby-lang.org/ja/latest/class/Hash.html
> ハッシュに含まれる要素の順序が保持されるようになりました。ハッシュにキーが追加された順序で列挙します。

guchi5 replied:
それをすると`create_matrix_for_long_option`が複雑になってしまうので、このままでいきたいと思います。

guchi5 replied:
`matrix`自体に左寄せ表示のフラグを格納することにしました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,66 @@ def show_files(matrix)
+  show_detail_files(create_matrix_for_long_option(file_attributes, COL_SIZE_FOR_DETAIL))
+  return
+end
+matrix = create_matrix(files, col_size)
```
こっちも `col_size` ではなく直接 `MAX_COL_SIZE` でいいように思いました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,66 @@ def show_files(matrix)
 return if files.empty?
 
-matrix = create_matrix(files, MAX_COL_SIZE)
+total_files_num = 0
```
`total_files_num` だとファイル数のように見えるので、 `total_file_blocks` とかにしたほうがいいかなと思いました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,66 @@ def show_files(matrix)
+  file_attributes = files.map do |file|
+    absolute_path = "#{File.expand_path(path, '.')}/#{file}"
+    total_files_num += File.stat(absolute_path).blocks / 2
+    status = File.stat(absolute_path).mode.to_s(8)
+    status = format('%06d', status.to_i)
```
前回も書きましたが、基本的に型が違うものを同じ変数にいれるようなやり方はバグを誘発しやすいので、避けたほうがいいです。そうすることに合理的な理由がない場合は避けましょう。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +85,66 @@ def show_files(matrix)
+def convert_file_mode(status, file)
+  absolute_path = "#{File.expand_path(ARGV[0]&.to_s || '.', '.')}/#{file}"
+  show_status = FILE_MODE_TABLE['file_type'][status[0..1].to_i].to_s
+  show_status += FILE_MODE_TABLE['permission'][status[3].to_i].to_s
+  show_status += FILE_MODE_TABLE['permission'][status[4].to_i].to_s
+  show_status += FILE_MODE_TABLE['permission'][status[5].to_i].to_s
```
この辺は繰り返し処理にしてしまってもよいと思います。

```ruby
(3..5).each do |i|
  show_status += FILE_MODE_TABLE['permission'][status[i].to_i].to_s
end
```

guchi5 replied:
rubocopからeachにするとだめだと言われてしまったので、このままでいきたいと思います。

yoshitsugu replied:
なるほど。`Perceived complexity` の警告がでますね。一旦大丈夫です。

yoshitsugu reviewed (commented):
コメントしました！

guchi5 committed:
可読性のための修正ver2

guchi5 committed:
左寄せ表示処理方法を修正

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -17,6 +47,33 @@ def create_matrix(files, max_col_size)
+  files.each_slice(row_size).with_index do |(*col), index|
+    valid_col = col.compact
+    max_size = valid_col.max_by(&:length).length
+    left_flag = true if [2, 3, 6].include?(index)
```
結局ここのロジックが 2, 3, 6 というパっと見わかりづらいインデックスに依存しているので、あまり可読性の向上に つながってないですね。
僕がイメージしてたのは、例えば、
https://github.com/guchi5/ruby-practices/pull/8/files#diff-b8915aacdb689607670be490d25d5de2ad9fb1362b301904d6b6089318a35af1R140
ここで返すのを配列ではなく、属性名をキーにしたHashにすれば属性名でアクセスできるようになるので可読性があがるのでは？という感じでした。

yoshitsugu reviewed (commented):
一点コメントしました。

guchi5 committed:
左寄せ表示処理方法を修正ver2

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -17,6 +47,33 @@ def create_matrix(files, max_col_size)
 
+def create_matrix_for_long_option(files, max_col_size)
+  matrix = []
+  col_name = %w[mode link user group file_size Date file]
```
ここの順番と
https://github.com/guchi5/ruby-practices/pull/8/files#diff-b8915aacdb689607670be490d25d5de2ad9fb1362b301904d6b6089318a35af1R140
ここの順番が離れたところで依存しているのはちょっとややこしく感じます。
`files` として渡ってくる引数側の順番がかわったり、新しい要素が追加されたりしたら、忘れずにこっちも変更しないといけないですよね。

https://github.com/guchi5/ruby-practices/pull/8/files#diff-b8915aacdb689607670be490d25d5de2ad9fb1362b301904d6b6089318a35af1R140
こちら側をHashにしたほうがいいかなと思うのですが、どうでしょうか。
```ruby
{ mode: show_status, link: hard_link_num, ... }
```

以前書きましたが、RubyのHashは順番を保持するので、順番を考慮する際も問題ないものの、もし明示的に順番を配列で与えたいということなら、定数として、

```ruby
FILE_STAT_DISPLAY_ORDER = %i[mode link user group file_size Date file]
```

みたいにして、その順番で呼ぶ、などは大丈夫だと思います。

guchi5 replied:
前者の方に変更しました！

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
表示用行列生成過程を変更

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -28,20 +87,63 @@ def show_files(matrix)
+  show_status += FILE_MODE_TABLE['permission'][status[3].to_i].to_s
+  show_status += FILE_MODE_TABLE['permission'][status[4].to_i].to_s
+  show_status += FILE_MODE_TABLE['permission'][status[5].to_i].to_s
+  special_permission = FILE_MODE_TABLE['special_permission'][status[2].to_i].to_s
+  special_permission = special_permission.upcase if show_status[3] != 'x'
```
ここのロジックがおかしいかもしれません。
組み込みで
```
-rwsr-sr-x 1 yoshitsugu yoshitsugu  0  2月  9  2023 test8.txt
```
のようになるファイルを渡すと
```
 -rwr-xr-x 1 yoshitsugu yoshitsugu  0 Feb  9 14:29 test8.txt
```
となってしまいました。

yoshitsugu reviewed (commented):
コメントしました。

guchi5 committed:
特殊権限表示処理を変更
- convert_file_modeの処理を分割する形で、update_special_permission関数を追加
- FILE_MODE_TABLEの特殊権限部分を変更

yoshitsugu reviewed (approved):
OKです！
